### ISSUE TITLE
How to implement support for Redis Sentinel?

### ISSUE THREAD

**Issue by Namoshek** [soketi/pws#10]
As I've written on Reddit already, it would be nice to have support for Redis Sentinel. But before we dig into details, a bit of background about what Redis Sentinel is:
> Redis Sentinel provides high availability for Redis. In practical terms this means that using Sentinel you can create a Redis deployment that resists without human intervention certain kinds of failures.
> 
> Redis Sentinel also provides other collateral tasks such as monitoring, notifications and acts as a configuration provider for clients.
> 
> This is the full list of Sentinel capabilities at a macroscopical level (i.e. the big picture):
> - Monitoring. Sentinel constantly checks if your master and replica instances are working as expected.
> - Notification. Sentinel can notify the system administrator, or other computer programs, via an API, that something is wrong with one of the monitored Redis instances.
> - Automatic failover. If a master is not working as expected, Sentinel can start a failover process where a replica is promoted to master, the other additional replicas are reconfigured to use the new master, and the applications using the Redis server are informed about the new address to use when connecting.
> - Configuration provider. Sentinel acts as a source of authority for clients service discovery: clients connect to Sentinels in order to ask for the address of the current Redis master responsible for a given service. If a failover occurs, Sentinels will report the new address.

In other words, Redis Sentinel ensures there is always a master available. In a failover scenario, this is ensured by promoting a replica to be the new master.

In terms of "connecting to Redis", we are essentially talking about a proxy which needs to be queried for its current master before we can connect to the actual Redis instance. Luckily, [`ioredis` has this built-in already](https://github.com/luin/ioredis#sentinel).

---

So my questions regarding a potential implementation are:

- Currently, the environment variables are basically the same options which are passed to `ioredis`. Do we want to use the same configuration style for Redis Sentinel as well? It would for sure be nice, but I see two problems here:
  - `ioredis` allows configuring multiple Sentinels, which seems hard/cumbersome to implement with statically mapped environment variables (`DB_REDIS_SENTINEL_0_HOST`, `DB_REDIS_SENTINEL_1_HOST`, ...).

  - There would be room for interpretation if both options (non-sentinel and sentinel) are configured. And if both were made optional, it would be a breaking change in my opinion.

- Would it be ok to have a `DB_REDIS_SENTINEL=true/false` setting to allow switching to Sentinel mode? We could then simply use a different connection configuration internally, based on the already existing environment variables `DB_REDIS_HOST` and `DB_REDIS_PORT`. This would not allow defining multiple Sentinels though. In my case, this is not an issue, because we are connecting to the Sentinels through a Kubernetes Service and a short outage of a few seconds is acceptable (and better for us than a complex configuration).

- Would it be ok to have a `DB_REDIS_URLS` option which allows to pass multiple `host:port` combinations in a comma separated fashion (`DB_REDIS_URLS=localhost:6379,localhost:6380`)? This solves some of the problems and introduces new ones (e.g. precedence of configuration options).

I think it is better to discuss this before starting an implementation. Eager to hear your feedback on this ideas!

**Comment by rennokki** [comment]
Thanks for the sentinel details. It seems like it's good to have, as long as Redis IO supports it out-of-the-box. I guess the other thing that needs to be implemented is the `RedisCluster` feature, but this can wait a bit.

[I've mentioned in some other issue](https://github.com/soketi/pws/issues/7#issuecomment-871911528) that the environment variables are JSON-ready, meaning that passing a JSON-string will transform it into the equivalent object. I think it's the best approach, but I still need to consider the comma-separated-values, they seem like a better approach 🤔

**Comment by Namoshek** [comment]
In that case I think it is best to stick with the simple approach and allow passing json. It binds the implementation a bit to `ioredis`, but for me that's fine since I do not expect this to change.

**Comment by rennokki** [comment]
Closed in #11

**Comment by rennokki** [comment]
@Namoshek Please tell me if the latest tag works well for you with Redis Sentinel.

**Comment by Namoshek** [comment]
@rennokki Works great locally, will give it a try in our dev Kubernetes environment as soon as possible I'm quite confident that it will work as expected! 👍

**Comment by rennokki** [comment]
Released 0.2.0 with the new features

**Comment by Namoshek** [comment]
Aaaand of course I forgot to answer here. 😄 

I've tested it over two weeks ago already and it is since running in our test environment without issues. I've also tested a failover with Redis Sentinel and `ioredis` ensures a connection is re-established. So, from my perspective, all seems fine with the update!
